### ISSUE TITLE
It's unsafe to retry merge when GitHub's pull request merge API returns a 500 error

### ISSUE THREAD

**Issue by styfle** [chdsbd/kodiak#397]
Kodiak has merged the same PR multiple times. It wasn't until I manually clicked the merge button that it stopped.

https://github.com/zeit/now/commits/master

https://github.com/zeit/now/pull/4412

![image](https://user-images.githubusercontent.com/229881/82328548-0c6c3a00-99ae-11ea-9cac-f349c6e04444.png)

**Comment by chdsbd** [comment]
This is concerning. I’m confused how this is even possible.

**Comment by chdsbd** [comment]
I'm assuming this is from a recent change to Kodiak and not a change in GitHub's API, so I've rolled back to the previously deployed version (c2c2690b4c1d0a1c4132acded6317941fd30cdb9) from (a10246d9d820331afcd65b206af4b4e3082ae551).

**Comment by chdsbd** [comment]
Okay, I think I found the problem. I think it's because the GitHub API was returning 500 errors, so Kodiak was retrying the merge api request (https://developer.github.com/v3/pulls/#merge-a-pull-request-merge-button).

<img width="843" alt="image" src="https://user-images.githubusercontent.com/1929960/82333532-d7afb100-99b4-11ea-8f3d-f7c2dfd464dd.png">

Kodiak currently retries on failure, but that obviously didn't work in this case. I need to think on this more about how to prevent this in the future. (any ideas are welcome).

It seems like the GitHub pull request merge api should be idempotent (you can only merge a PR once, or so I thought), but that's obviously not the case when GitHub's API is failing.

I'm sorry for how Kodiak ran wild on your repository.

**Comment by chdsbd** [comment]
FYI, the rollback will not have fixed this. It could still happen depending on GItHub's API stability.

Tonight I'll try and change Kodiak's behavior so on merge failure we add some label, `kodiak:merge-failed` instead of retrying.

**Comment by styfle** [comment]
Thanks for the quick response!

> It seems like the GitHub pull request merge api should be idempotent (you can only merge a PR once, or so I thought),

Agreed, It might be worth reaching out to the GitHub staff on this one.

**Comment by styfle** [comment]
> Tonight I'll try and change Kodiak's behavior so on merge failure we add some label, kodiak:merge-failed, instead of retrying.

Is a label necessary? I think updating the github check with text that says the merged failed would be sufficient. This is already the place I look to find out why a PR is not merged yet (checks missing, queued, etc).

**Comment by sbdchd** [comment]
Possibly related to the 500s, GitHub's having an incident https://www.githubstatus.com/incidents/6hjb1189t0y9

**Comment by chdsbd** [comment]
@styfle I've opened a GitHub support issue and when I hear back (usually takes two weeks) I'll post the response here.

@styfle My thought is that with the label a user could reenable Kodiak by removing it. I would also update the status check saying the merge failed.

**Comment by styfle** [comment]
@chdsbd I think you could make it a little smarter and instead remove the "automerge" label (or whatever was configured as `automerge_label`). 

For example: https://github.com/zeit/now/blob/master/.kodiak.toml#L4

In the case there is no automerge label configured and kodiak automerges always, then you could add the merge failed label as a fallback.

**Comment by chdsbd** [comment]
@styfle Good idea. I'll do that and I think I'll probably add a comment explaining the problem, like Kodiak does with merge conflicts when the auto merge label is enabled.

**Comment by chdsbd** [comment]
@styfle I've opened a pull request (#398) to address this issue.

I realized while making the PR that we probably don't want to depend on the automerge label because even if the automerge label is removed Kodiak could still update or approve the pull request depending on the user's Kodiak configuration.

Instead the pull request adds a new label that disables Kodiak from making any pull request update, comment, label change, approval, or merge (Kodiak will still update status checks). This change is more general and I think it could be useful for another feature I've considered (merge timeouts).
